### PR TITLE
Use finalizers for RBAC cleanup

### LINKED ISSUE #87: Fix cross-namespace owner references for RBAC resources
#### What is the current behavior?

As part of https://github.com/containership/cluster-manager/issues/85, we realized that we're probably not correctly following the supposed [limitations of owner references](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#owners-and-dependents) for resources.  See the discussion on #85 for more details.

#### What is the expected behavior?

Fully understand the limitations of ownerRefs and if we are not following them properly, update the code to either use a good ownerRef or manually handle what the GC was handling for us.

#### Additional Context
See #85.

### PR DESCRIPTION
**Pull request by mattkelly**
#### What does this pull request accomplish?

Use finalizers for RBAC cleanup

Previously, we were relying on Kubernetes garbage collection. This should
have never worked due to cross-namespace owner references but was working
due to a bug in  Kubernetes.

See https://github.com/containership/cluster-manager/issues/87 for more
details.

 #### What issue(s) does this fix?
* Fixes #87

 #### Additional Considerations

 ### Testing

 #### Setup

- Build and push image to `mattkelly/cloud-coordinator:latest`
- Provision a 1.15.2 cluster
- Update coordinator image

 #### Instructions

- See that existing CRs are recreated with finalizers
- Delete CR and see that roles/bindings are cleaned up before delete completes (and see that CR is eventually recreated)
- Delete role binding for cluster in UI and see that binding is cleaned up before CR delete completes

 ### Dependencies
 None

### CODE REVIEW

File: pkg/apis/auth.containership.io/v3/zz_generated.deepcopy.go [hunk 1]
```diff
@@ -82,7 +82,7 @@ func (in *AuthorizationRoleBinding) DeepCopyObject() runtime.Object {
 func (in *AuthorizationRoleBindingList) DeepCopyInto(out *AuthorizationRoleBindingList) {
 	*out = *in
 	out.TypeMeta = in.TypeMeta
-	in.ListMeta.DeepCopyInto(&out.ListMeta)
+	out.ListMeta = in.ListMeta
```
Comment (mattkelly) [inline pkg/apis/auth.containership.io/v3/zz_generated.deepcopy.go]: Code generation was out of date for some reason. Had this in a separate commit but accidentally squashed it in 🤦‍♂

**Comment by mattkelly** [comment]
I'd like to do some more thorough testing before this is merged, but it's good for an initial review. May also refactor a bit.

File: pkg/apis/auth.containership.io/v3/zz_generated.deepcopy.go (hunk 1 above)
Comment (mattkelly) [reply pkg/apis/auth.containership.io/v3/zz_generated.deepcopy.go]: Actually, my local dependencies were just messed up due to go module experimentation. Deleted go mod stuff and reran `dep ensure` and it's good to go.

File: pkg/coordinator/authorization_role_binding_controller.go [hunk 2]
```diff
@@ -84,7 +84,7 @@ func NewAuthorizationRoleBindingController(kubeclientset kubernetes.Interface,
 			// unexpected *Binding deletions.
 			c.enqueueAuthorizationRoleBinding(new)
 		},
-		// We don't care about deletes because GC will take care of orphaned ClusterRoleBindings
+		DeleteFunc: c.enqueueAuthorizationRoleBinding,
```
Comment (mattkelly) [inline pkg/coordinator/authorization_role_binding_controller.go]: We don't need to listen for deletes

File: pkg/coordinator/authorization_role_controller.go [hunk 3]
```diff
@@ -80,7 +80,7 @@ func NewAuthorizationRoleController(kubeclientset kubernetes.Interface,
 			// unexpected ClusterRole deletions.
 			c.enqueueAuthorizationRole(new)
 		},
-		// We don't care about deletes because GC will take care of orphaned ClusterRoles
+		DeleteFunc: c.enqueueAuthorizationRole,
```
Comment (mattkelly) [inline pkg/coordinator/authorization_role_controller.go]: We don't need to listen to deletes

**Comment by nicktate** [comment]
Nice work overall 👏  I will await any further refactors before doing a final review but from a high level it makes sense and I think is a nice approach to solving this issue with GC!

**Comment by mattkelly** [comment]
This is fully tested, refactored a bit, and ready for final review.

File: pkg/coordinator/authorization_role_binding_controller.go [hunk 4]
```diff
@@ -235,7 +270,8 @@ func (c *AuthorizationRoleBindingController) authorizationRoleBindingSyncHandler
 			return errors.Wrapf(err, "getting ClusterRoleBinding %s for reconciliation", id)
 		}
 
-		log.Debugf("%s: Updating existing ClusterRoleBinding %s", authorizationRoleBindingControllerName, id)
+		// Always call update and let Kubernetes figure out if an update is
+		// actually needed instead of determining that ourselves
 		binding := clusterRoleBindingFromAuthorizationRoleBinding(*authorizationRoleBinding)
```
[the remark this answers is not in the document]
Comment (mattkelly) [reply pkg/coordinator/authorization_role_binding_controller.go]: We exit from this function immediately if the CR no longer exists. If the CR is marked for deletion, we run the finalizer logic and return early. I don't think this should be a concern unless I'm misunderstanding.

File: pkg/coordinator/authorization_role_binding_controller.go [hunk 5]
```diff
@@ -253,22 +289,51 @@ func (c *AuthorizationRoleBindingController) authorizationRoleBindingSyncHandler
 			return errors.Wrapf(err, "getting RoleBinding %s for reconciliation in namespace %s", id, roleNamespace)
 		}
 
-		log.Debugf("%s: Updating existing RoleBinding %s", authorizationRoleBindingControllerName, id)
+		// Always call update and let Kubernetes figure out if an update is
+		// actually needed instead of determining that ourselves
 		binding := roleBindingFromAuthorizationRoleBinding(*authorizationRoleBinding)
```
[the remark this answers is not in the document]
Comment (mattkelly) [reply pkg/coordinator/authorization_role_binding_controller.go]: See https://github.com/containership/cluster-manager/pull/92/files#r315777322

File: pkg/coordinator/authorization_role_controller.go [hunk 6]
```diff
@@ -224,12 +255,27 @@ func (c *AuthorizationRoleController) authorizationRoleSyncHandler(key string) e
 		return errors.Wrapf(err, "getting ClusterRole %s for reconciliation", id)
 	}
 
-	log.Debugf("%s: Updating existing ClusterRole %s", authorizationRoleControllerName, id)
+	// Always call update and let Kubernetes figure out if an update is
+	// actually needed instead of determining that ourselves
 	role := clusterRoleFromAuthorizationRole(*authorizationRole)
 	_, err = c.kubeclientset.RbacV1().ClusterRoles().Update(&role)
```
[the remark this answers is not in the document]
Comment (mattkelly) [reply pkg/coordinator/authorization_role_controller.go]: See https://github.com/containership/cluster-manager/pull/92/files#r315777322

File: pkg/coordinator/authorization_role_binding_controller.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nicktate) [reply pkg/coordinator/authorization_role_binding_controller.go]: Nope that clarifies it for me, I missed that `return nil` line above

**Pull request merged**